### PR TITLE
Add PortraitNet.

### PR DESCRIPTION
**Pull request by txyugood**
1. Add portraitNet model.
2. Add RandomAffine in transform.py for data augmentation.
3. Add kl_loss and focal_loss for PortraintNet.
4. Modify train.py for PortraintNet.
5. Add EG1800 and Supervisely dataset.
6. Add stepdecay in config.py for training strategy.

### CODE REVIEW

File: paddleseg/transforms/transforms.py [hunk 1]
```diff
@@ -932,3 +910,87 @@ def __call__(self, im, label=None):
             return (im, )
         else:
             return (im, label)
+
+
+@manager.TRANSFORMS.add_component
+class RandomAffine:
```
Comment (LutaoChu) [inline paddleseg/transforms/transforms.py:1089]: 这个指缩放+旋转吧？是否可以通过已有的StepScaling+Rotation组合就好了，不用另外开发？

File: paddleseg/datasets/eg1800.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+from paddleseg.transforms import Compose
+import paddleseg.transforms.functional as F
+import cv2
+import numpy as np
+import random
+from PIL import Image, ImageEnhance
+import copy
+
+
+def padding(img_ori, mask_ori, size=224, padding_color=128):
```
Comment (LutaoChu) [inline paddleseg/datasets/eg1800.py]: 这些transform之前不是说可以复用的复用原来的transform，没有的新增到transform.py中吗？

File: paddleseg/datasets/supervisely.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+import os
+
+from paddleseg.datasets import EG1800
+from paddleseg.cvlibs import manager
+from paddleseg.transforms import Compose
+
+
+@manager.DATASETS.add_component
+class SUPERVISELY(EG1800):
```
Comment (LutaoChu) [inline paddleseg/datasets/supervisely.py]: 尽量继承DataSet吧，数据集间解耦

File: paddleseg/datasets/eg1800.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        pass
+    '''
+    cv2.drawContours(myImg, countours, -1, 1, 10)
+    diff = mask + myImg
+    diff[diff < 2] = 0
+    diff[diff == 2] = 1
+    return diff
+    '''
+    cv2.drawContours(myImg, countours, -1, 1, 4)
+    # myImg[myImg==1] = 255
```
Comment (LutaoChu) [inline paddleseg/datasets/eg1800.py]: 删掉注释

File: paddleseg/models/portraitnet.py [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+        if self.residual is not None:
+            residual = self.residual(x)
+
+        out += residual
+        out = self.relu(out)
+        return out
+
+
+@manager.MODELS.add_component
+class MobileNetV2(nn.Layer):
```
Comment (LutaoChu) [inline paddleseg/models/portraitnet.py]: 准确的讲，mobilenetv2是作为portraitnet的backbone吧
建议单独放一个文件，和mobilnetv3一样，放到paddleseg/models/backbone下

File: paddleseg/models/portraitnet.py [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+        self.stage6 = nn.Sequential(  # 1/32
+            InvertedResidual(self.depth(96), self.depth(160), 2, 6),
+            InvertedResidual(self.depth(160), self.depth(160), 1, 6),
+            InvertedResidual(self.depth(160), self.depth(160), 1, 6),
+        )
+
+        self.stage7 = InvertedResidual(self.depth(160), self.depth(320), 1,
+                                       6)  # 1/32
+
+        self.deconv1 = nn.Conv2DTranspose(self.depth(96),
```
Comment (LutaoChu) [inline paddleseg/models/portraitnet.py]: decode的部分放到portraitnet中

File: paddleseg/models/portraitnet.py [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+                    sublayer._kernel_size[0], sublayer.weight.shape[0],
+                    sublayer.weight.shape[1])  # same as caffe
+                initializer = nn.initializer.Assign(initial_weight)
+                initializer(sublayer.weight, sublayer.weight.block)
+        if self.pretrained is not None:
+            utils.load_entire_model(self, self.pretrained)
+
+
+@manager.MODELS.add_component
+class PortraitNet(nn.Layer):
```
Comment (LutaoChu) [inline paddleseg/models/portraitnet.py]: 这段放在文件的前面吧

**Review by LutaoChu — CHANGES_REQUESTED** [review]

File: paddleseg/transforms/transforms.py (hunk 1 above)
Comment (txyugood) [reply paddleseg/transforms/transforms.py:1089]: 这个是平移、缩放和旋转。我之前试过通过StepScaling+Rotation但是效果完全不一样。所以按照论文实现了RandomAffine。

File: paddleseg/datasets/eg1800.py (hunk 2 above)
Comment (txyugood) [reply paddleseg/datasets/eg1800.py]: 这个padding跟transform中的padding略微有点区别，这个是居中的，transform里面的是是靠左的，而且不能设置。那我修改一下transform里的padding?

File: paddleseg/datasets/supervisely.py (hunk 3 above)
Comment (txyugood) [reply paddleseg/datasets/supervisely.py]: 好的，我改下这个。

**Review by LutaoChu — DISMISSED** [review]

File: paddleseg/core/train.py [hunk 8]
```diff
@@ -104,8 +109,10 @@ def train(model,
             optimizer)  # The return is Fleet object
         ddp_model = paddle.distributed.fleet.distributed_model(model)
 
-    batch_sampler = paddle.io.DistributedBatchSampler(
-        train_dataset, batch_size=batch_size, shuffle=True, drop_last=True)
```
Comment (LutaoChu) [inline paddleseg/core/train.py]: yapf版本改为0.26.0，所有文件重新格式化一下吧

File: paddleseg/datasets/eg1800.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (LutaoChu) [inline paddleseg/datasets/eg1800.py]: 所有文件license里的2020改成2021

File: paddleseg/datasets/eg1800.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+import os
+
+from paddleseg.datasets import Dataset
+from paddleseg.cvlibs import manager
+from paddleseg.transforms import Compose
+import paddleseg.transforms.functional as F
+import cv2
+import numpy as np
+import copy
```
Comment (LutaoChu) [inline paddleseg/datasets/eg1800.py]: 所有文件的import顺序应该符合PEP8规范

File: configs/portraitnet/portraitnet_eg8000_224x224_46k.yml [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+      mean: [0.485, 0.458, 0.408]
+      std: [0.23, 0.23, 0.23]
+  transforms2: null
+  mode: val
+
+optimizer:
+  type: adam
+  weight_decay: 5.0e-4
+
+learning_rate:
```
Comment (LutaoChu) [inline configs/portraitnet/portraitnet_eg8000_224x224_46k.yml]: 最新用lr_scheduler，learning_rate要被废弃了

File: paddleseg/datasets/eg1800.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+import paddleseg.transforms.functional as F
+
+
+@manager.DATASETS.add_component
+class EG1800(Dataset):
+    """
+    EG1800 dataset `http://xiaoyongshen.me/webpage_portrait/index.html`.
+
+    Args:
+        transforms (list): A list of image transformations.
```
Comment (LutaoChu) [inline paddleseg/datasets/eg1800.py]: 缺少transforms2的注释

File: paddleseg/datasets/supervisely.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+        if self.mode == "val":
+            im, label = self.transforms(im=im, label=label)
+            im = np.float32(im[::-1, :, :])
+            im_aug = copy.deepcopy(im)
+        else:
+            im, label = self.transforms(im=im, label=label)
+            # add augmentation
+            std = np.array(self.std)[:, np.newaxis, np.newaxis]
+            mean = np.array(self.mean)[:, np.newaxis, np.newaxis]
+            im_aug = im * std + mean
```
Comment (LutaoChu) [inline paddleseg/datasets/supervisely.py]: 这个地方逻辑有点别扭。最好是经过共同的transform后，出来两个分支transform，分别处理im和im_aug。
EG1800也是一样

File: paddleseg/models/losses/focal_loss.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+        ignore_index (int64): Specifies a target value that is ignored
+            and does not contribute to the input gradient. Default ``255``.
+    """
+
+    def __init__(self, gamma=2.0, ignore_index=255):
+        super(FocalLoss, self).__init__()
+        self.gamma = gamma
+        self.ignore_index = ignore_index
+
+    def forward(self, inp, target):
```
Comment (LutaoChu) [inline paddleseg/models/losses/focal_loss.py]: inp, target改成logit，label吧

File: paddleseg/models/backbones/mobilenetv2.py [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+            return self.conv(x)
+
+
+@manager.MODELS.add_component
+class MobileNetV2(nn.Layer):
+    """
+        The MobileNetV2 implementation based on PaddlePaddle.
+
+        The original article refers to
+        Song-Hai Zhanga, Xin Donga, Jia Lib, Ruilong Lia, Yong-Liang Yangc
```
Comment (LutaoChu) [inline paddleseg/models/backbones/mobilenetv2.py]: Reference改成mobilenetv2

File: paddleseg/models/losses/focal_loss.py [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+import paddle.nn.functional as F
+
+from paddleseg.cvlibs import manager
+
+
+@manager.LOSSES.add_component
+class FocalLoss(nn.Layer):
+    """
+    The Focal Loss implement of Portrait Net.
+
```
Comment (LutaoChu) [inline paddleseg/models/losses/focal_loss.py:30]: 改成
'''
Focal Loss.

Code referenced from: 
https://github.com/clcarwin/focal_loss_pytorch/blob/master/focalloss.py
'''

File: paddleseg/models/losses/focal_loss.py [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+        inp = paddle.reshape(inp, [-1, 2])
+        target = paddle.reshape(target, [-1, 1])
+        range_ = paddle.arange(0, target.shape[0])
+        range_ = paddle.unsqueeze(range_, axis=-1)
+        target = paddle.cast(target, dtype='int64')
+        target = paddle.concat([range_, target], axis=-1)
+        logpt = F.log_softmax(inp)
+        logpt = paddle.gather_nd(logpt, target)
+
+        pt = paddle.to_tensor(np.exp(logpt.numpy()))
```
Comment (LutaoChu) [inline paddleseg/models/losses/focal_loss.py]: 为什么不用paddle.exp呢？

File: paddleseg/models/losses/focal_loss.py [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+
+    def __init__(self, gamma=2.0, ignore_index=255):
+        super(FocalLoss, self).__init__()
+        self.gamma = gamma
+        self.ignore_index = ignore_index
+
+    def forward(self, inp, target):
+        inp = paddle.reshape(inp, [inp.shape[0], inp.shape[1], -1])
+        inp = paddle.transpose(inp, [0, 2, 1])
+        inp = paddle.reshape(inp, [-1, 2])
```
Comment (LutaoChu) [inline paddleseg/models/losses/focal_loss.py]: 应该是inp = paddle.reshape(inp, [-1, inp.shape[2])吧，用的时候不一定是二分类

File: paddleseg/models/losses/focal_loss.py [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+            and does not contribute to the input gradient. Default ``255``.
+    """
+
+    def __init__(self, gamma=2.0, ignore_index=255):
+        super(FocalLoss, self).__init__()
+        self.gamma = gamma
+        self.ignore_index = ignore_index
+
+    def forward(self, inp, target):
+        inp = paddle.reshape(inp, [inp.shape[0], inp.shape[1], -1])
```
Comment (LutaoChu) [inline paddleseg/models/losses/focal_loss.py]: 加上注释   ```# N,C,H,W => N,C,H*W```

File: paddleseg/models/losses/focal_loss.py [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+        inp = paddle.transpose(inp, [0, 2, 1])
```
Comment (LutaoChu) [inline paddleseg/models/losses/focal_loss.py]: 加上注释 ```# N,C,H*W => N,H*W,C```

File: paddleseg/models/losses/focal_loss.py (hunk 18 above)
Comment (LutaoChu) [reply paddleseg/models/losses/focal_loss.py]: 加上注释  ```# N,H*W,C => N*H*W,C```

File: paddleseg/models/losses/kl_loss.py [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+        temperature (float): the coefficient of kl_loss.
+    """
+
+    def __init__(self, ignore_index=255, temperature=1):
+        super(KLLoss, self).__init__()
+        self.ignore_index = ignore_index
+        self.kl_loss = nn.KLDivLoss(reduction="mean")
+        self.temperature = temperature
+
+    def forward(self, inp, target):
```
Comment (LutaoChu) [inline paddleseg/models/losses/kl_loss.py]: inp, target改成logit，label

File: paddleseg/models/losses/kl_loss.py [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+import paddle.nn as nn
+import paddle.nn.functional as F
+
+from paddleseg.cvlibs import manager
+
+
+@manager.LOSSES.add_component
+class KLLoss(nn.Layer):
+    """
+    The Kullback-Leibler divergence Loss implement of Portrait Net.
```
Comment (LutaoChu) [inline paddleseg/models/losses/kl_loss.py]: 去掉portraitnet
加上
```
    Code referenced from: 
    https://github.com/peterliht/knowledge-distillation-pytorch/blob/master/model/net.py
    
    Compute the knowledge-distillation (KD) loss given outputs, labels.
    "Hyperparameters": temperature and alpha
```

**Review by LutaoChu — CHANGES_REQUESTED** [review]

File: paddleseg/models/losses/focal_loss.py [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+    """
+    The Focal Loss implement of Portrait Net.
+
+    Args:
+        gamma (float): the coefficient of Focal Loss.
+        ignore_index (int64): Specifies a target value that is ignored
+            and does not contribute to the input gradient. Default ``255``.
+    """
+
+    def __init__(self, gamma=2.0, ignore_index=255):
```
Comment (LutaoChu) [inline paddleseg/models/losses/focal_loss.py]: 为了能让focal loss更通用，加上edge_label=False参数，通过edge_label判断是不是用edge监督

File: paddleseg/models/losses/focal_loss.py [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+        super(FocalLoss, self).__init__()
+        self.gamma = gamma
+        self.ignore_index = ignore_index
```
Comment (LutaoChu) [inline paddleseg/models/losses/focal_loss.py:40]: 加上self.edge_label = edge_label

File: paddleseg/core/train.py [hunk 25]
```diff
@@ -42,6 +42,11 @@ def loss_computation(logits_list, labels, losses, edges=None):
         # Whether to use edges as labels According to loss type.
         if loss_i.__class__.__name__ in ('BCELoss', ) and loss_i.edge_label:
```
Comment (LutaoChu) [inline paddleseg/core/train.py]: 改成 if loss_i.__class__.__name__ in ('BCELoss',  "FocalLoss") and loss_i.edge_label:

File: paddleseg/core/train.py [hunk 26]
```diff
@@ -42,6 +42,11 @@ def loss_computation(logits_list, labels, losses, edges=None):
...
             loss_list.append(losses['coef'][i] * loss_i(logits, edges))
+        elif loss_i.__class__.__name__ in ("FocalLoss", ):
+            loss_list.append(losses['coef'][i] * loss_i(logits, edges))
```
Comment (LutaoChu) [inline paddleseg/core/train.py:45]: 去掉上面两行

**Review by LutaoChu — CHANGES_REQUESTED** [review]

File: paddleseg/models/portraitnet.py [hunk 27]
```diff
@@ -0,0 +1,215 @@
...
+            logits_ori_list[1]
+        ]
+
+    def init_weight(self):
+        if self.pretrained is not None:
+            utils.load_entire_model(self, self.pretrained)
+
+
+class PortraitNetHead(nn.Layer):
+    def __init__(self, num_classes, add_edge=False):
```
Comment (LutaoChu) [inline paddleseg/models/portraitnet.py]: 开放参数min_channel，channel_ratio吧

File: paddleseg/models/portraitnet.py [hunk 28]
```diff
@@ -0,0 +1,215 @@
...
+    (https://www.yongliangyang.net/docs/mobilePotrait_c&g19.pdf).
+
+    Args:
+        num_classes (int, optional): The unique number of target classes.  Default: 2.
+        backbone (Paddle.nn.Layer): Backbone network, currently support MobileNetV2.
+        add_edge (bool, optional): Whether output to edge. Default: False
+        pretrained (str, optional): The path or url of pretrained model. Default: None
+    """
+
+    def __init__(self, num_classes, backbone, add_edge=False, pretrained=None):
```
Comment (LutaoChu) [inline paddleseg/models/portraitnet.py]: 开放参数min_channel，channel_ratio吧

File: paddleseg/models/backbones/mobilenetv2.py [hunk 29]
```diff
@@ -0,0 +1,168 @@
...
+
+    def forward(self, x):
+        if self.use_res_connect:
+            return x + self.conv(x)
+        else:
+            return self.conv(x)
+
+
+@manager.MODELS.add_component
+class MobileNetV2(nn.Layer):
```
Comment (LutaoChu) [inline paddleseg/models/backbones/mobilenetv2.py:22]: 提到文件最前头吧

**Review by LutaoChu — DISMISSED** [review]

**Review by michaelowenliu — APPROVED** [review]
